virtru-audit-client/__main__.py

```diff
@@ -0,0 +1,105 @@
+                    dest='configFile',
+                    required=True)
+parser.add_argument('-s', '--start',
+                    help='Start date of query.  Example: -s 2019-01-01',
+                    dest='startDate',
+                    default='2010-01-01',
+                    required=False)
+parser.add_argument('-e', '--end',
+                    help='End date of query.  Example: -e 2019-02-01',
+                    dest='endDate',
+                    default='2100-01-01',
+                    required=False)
+parser.add_argument('--csv',
+                    help='CSV output folder.  If defined CSV will be exported',
+                    dest='csv',
+                    default=None,
+                    required=False)
+parser.add_argument('--json',
+                    help='Json output folder.  If defined Json will be exported.  Example: output/  '
+                    '/home/user/json/',
+                    dest='json',
+                    default=None,
+                    required=False)
+parser.add_argument('--sysloghost',
+                    help='Syslog server.  If defined syslog will be exported',
+                    dest='sysloghost',
+                    default=None,
+                    required=False)
+parser.add_argument('--syslogport',
+                    help='Syslog port.  If a different port is required.',
+                    dest='syslogport',
+                    default=None,
+                    required=False)
+
+args = parser.parse_args()
+
+config = utils.getConfig(args.configFile)
+apiTokenId = config['apiTokenId']
+apiTokenSecret = config['apiTokenSecret']
+apiHost = config['apiHost']
+apiPath = config['apiPath']
+
+bookMark = utils.getNextPageStartKey()
+nextPageStartKey = None if not bookMark else bookMark['nextpagestartkey']
+
+
+queryStart = args.startDate
+queryEnd = args.endDate
+jsonFolderPath = args.json
+csvFolderPath = args.csv
+syslogHost = args.sysloghost
+syslogPort = args.syslogport
+
+auditClient = AuditClient(apiTokenSecret, apiTokenId,
+                          apiHost, apiPath)
+
+req = {
+    'method': 'GET',
+    'query': {
+        'start': queryStart,
+        'end': queryEnd
+    }
+}
+
+if(nextPageStartKey):
+    req['query']['nextPageStartKey'] = nextPageStartKey
+
+hasMore = True
+iteration = 1
+
+while hasMore:
+    try:
+        records = auditClient.fetchRecords(req)
+        if(jsonFolderPath):
+            utils.exportToJson(jsonFolderPath, records['docs'])
+        if(csvFolderPath):
+            utils.exportToCsv(csvFolderPath, records['docs'])
+        if(syslogHost is not None and syslogPort is not None):
+            utils.exportToSysLog(syslogHost, syslogPort, records['docs'])
+
+        if 'nextPageStartKey' in records:
+            req['query']['nextPageStartKey'] = records['nextPageStartKey']
+            nextPageStartKey = records['nextPageStartKey']
+        else:
+            hasMore = False
+            if nextPageStartKey:
```

Comment:
If you're in this `else` then there is not another `nextPageStartKey`, so why would you want to check this condition?

Reply:
I think what this logic is trying to accomplish is to save the bookmark from the last request (set on line 95) in a file to be used on the next invocation of the script. However that doesn't totally make sense to me because it still makes that one extra request after the last bookmark and writes the resulting records which means the next time the script is run there is still going to be overlap. Can you weigh in @chadtsigler ?